compiler_opt/rl/corpus.py

```diff
@@ -12,31 +12,145 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""ModuleSpec definition and utility command line parsing functions."""
+"""Corpus and related concepts."""
+import abc
 import math
 import random
 import re
 
 from absl import logging
 from dataclasses import dataclass
-from typing import Any, Dict, Iterable, List, Optional, Tuple
+from typing import Any, Dict, List, Optional, Tuple
 
 import json
 import os
+import tensorflow as tf
 
 from compiler_opt.rl import constant
 
+# Alias to better self-document APIs. Represents a complete, ready to use
+# command line, where all the flags reference existing, local files.
+FullyQualifiedCmdLine = Tuple[str, ...]
+
+
+def _localize_cmdline(
+    orig_options: Tuple[str, ...],
+    context: Any = None,
+    additional_flags: Tuple[str, ...] = (),
+    delete_flags: Tuple[str, ...] = (),
+    replace_flags: Optional[Dict[str, str]] = None) -> List[str]:
+  option_iterator = iter(orig_options)
+  matched_replace_flags = set()
+  replace_flags = replace_flags if replace_flags is not None else {}
+
+  option = next(option_iterator, None)
+  cmdline = []
+  while option is not None:
+    if any(option.startswith(flag) for flag in delete_flags):
+      if '=' not in option:
+        next(option_iterator, None)
+    else:
+      matching_replace = [
+          flag for flag in replace_flags if option.startswith(flag)
+      ]
+      if not matching_replace:
+        cmdline.append(option)
+      else:
+        assert len(matching_replace) == 1
+        flag = matching_replace[0]
+        if flag in matched_replace_flags:
+          raise ValueError(f'{flag} was matched twice')
+        matched_replace_flags.add(flag)
+
+        if '=' not in option:
+          next(option_iterator, None)
+          cmdline.extend([option, replace_flags[flag].format(context=context)])
+        else:
+          cmdline.append(flag + '=' +
+                         replace_flags[flag].format(context=context))
+
+    option = next(option_iterator, None)
+  if len(matched_replace_flags) != len(replace_flags):
+    raise ValueError('flags that were expected to be replaced were not found')
+  cmdline.extend([f.format(context=context) for f in additional_flags])
+  return cmdline
+
+
+@dataclass(frozen=True)
+class LoadedModuleSpec:
+  """Encapsulates the loaded data of a module and the rules to persist it.
+
+  A LoadedModuleSpec can be passed to a remote location. There, given a local
+  directory, to_module_spec can be called, resulting in the data being saved
+  under that directory, the final compiler command line fully computed, and a
+  ready-to-use FullyQualifiedCmdLine returned.
+  """
+  name: str
+  module: bytes
+  thinlto_index: Optional[bytes] = None
```

Comment:
nit:

module->loaded_ir
thinlto_index->loaded_thinlto_index

to emphasize that they are loaded bytes instead of paths

Reply:
good point, done